---
title: "POULD: Phased Or Unphased Linkage Disequilibrium"
author: "Steven J. Mack, PhD -- sjmack@chori.org"
date: "July 3, 2019"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using POULD}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
* Package Version: 0.8.3.9000

# Overview

The *pould* package (pronounced "pooled") calculates four linkage disequilibrium (LD) statistics -- *D<sup>'</sup>*, *W<sub>n</sub>* and the conditional asymmetric LD (cALD) measures, *W<sub>A/B</sub>* and *W<sub>B/A</sub>*,  for genotype data from pairs of genetic loci, and can treat these data as either phased or unphased for these calculations. 

The package includes a wrapper function that parses multi-locus haplotypes in the [17th International HLA and Immunogenetics Workshop](http://ihiws.org) (IHIW) Family Haplotype Project's HaplObserve output format. This wrapper function generates output files in the R working directory.

The package also includes a function that applies a sign test to LD values for phased and unphased haplotypes generated by the wrapper function for an IHIW Family Haplotype Project-formatted dataset, as well as a function that generates heat-maps for each LD measure. For more information, see:<br>
Osoegawa et. al. [Hum Immunol. In Press, 2019](https://doi.org/10.1016/j.humimm.2019.01.010).

For information about these LD measures, see:<br>
Hedrick PW. [Genetics; 1987,117,331-41](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1203208/pdf/331.pdf).<br>
Cram√©r H. [Mathematical Models of Statistics](https://archive.org/details/in.ernet.dli.2015.13917), 1946, Princeton University Press, Princeton NJ.<br>
Thomson G, Single RM. [Genetics. 2014;198(1):321-31](https://doi.org/10.1534/genetics.114.165266).

*Pould* accepts genotype and haplotype data for *individual subjects* as input data. To calculate the cALD measures using haplotype frequency data, try the [*asymLD*](https://cran.r-project.org/web/packages/asymLD/index.html) package. See: Single et al. [Hum Immunol. 2016;77(3):288-294](https://doi.org/10.1016/j.humimm.2015.09.001) for more information about *asymLD*. 

**Note**: *Pould*'s *cALD()* function calls the *LDkl()* function from the R *gap* package. The version of the *gap* package currently in CRAN (v1.1-22) limits *LDkl()* to 100 alleles per locus. Analyses in which the product of the number of alleles at the two loci is more than 10000 will cause R to abort. This limit has been changed to 1000 alleles per locus in *gap* version 1.1-23. This version is available on GitHub, and can be installed using the *devtools* package -- `devtools::install_github("jinghuazhao/R/gap")`.

## Functions and Input Formats

### cALD

The *cALD()* function calculates the LD measures from genotype data for pairs of loci. The `drb1.dqb1.demo` dataset, shown below, illustrates the input format. The function accepts a 4-column data frame or  tab-delimited text file with data for the first locus in columns 1 and 2, and data for the second locus in columns 3 and 4. 

```
> drb1.dqb1.demo[1:6,]

##    DRB1  DRB1  DQB1  DQB1
## 1 15:01 07:01 06:02 03:03
## 2 04:05 13:01 03:02 06:03
## 3 04:01 13:02 03:02 06:09
## 4 16:02 04:05 05:02 03:02
## 5 15:01 07:01 06:02 02:02
## 6 04:01 04:01 03:02 03:02
```
The locus names should be used as column headers, with one allele/variant in each column. Each row represents a single subject. The headers for columns 1 and 2 must be identical, as must the headers for columns 3 and 4. If phase is known, *cALD()* assumes that columns 1 and 3 represent one haplotype, and that columns 2 and 4 represent the second haplotype. 

While HLA data are shown above, *cALD()* can accept any genetic data. The example below combines data for HLA-DQA1 and rs7743506.
```
      HLA-DQA1       HLA-DQA1     rs7743506     rs7743506
        02:01          01:02           C             C
        04:01          01:02           A             C
        04:01          05:01           A             C
        01:01          01:02           C             C
        04:01          01:01           A             C
        05:01          01:02           C             C
```

### LDWrap

The *LDWrap()* function parses haplotype data formatted using the 17th International HLA and Immunogenetic Workshop Family Haplotype Project's GL String-based format. The function accepts a data frame or CSV formatted text file, and passes genotype data for all pairs of loci in that dataset to *cALD()* for LD analysis.

A minimal *LDWrap()* data file or data frame contains two columns named "Relation" and "Gl String". Other columns are allowed, but are ignored. The `hla.hap.demo` dataset (shown below in edited form) illustrates the input format. Each row contains data for a single subject. The "Relation"" column can contain any text string; however, values such as "mother", "father" and "child" are standard for the Family HLA Data Project. *LDWrap()* will ignore all rows in which `Relation=child`; rows with any other value in the "Relation" column will be processed. 
```
Relation	Gl String
Subject	HLA-A*02:01~HLA-C*07:02~HLA-B*07:02+HLA-A*01:01~HLA-C*06:02~HLA-B*57:01
Subject	HLA-A*03:01~HLA-C*07:01~HLA-B*49:01+HLA-A*01:01~HLA-C*07:01~HLA-B*08:01
Subject	HLA-A*11:01~HLA-C*04:01~HLA-B*15:01+HLA-A*03:01~HLA-C*08:02~HLA-B*14:02
Subject	HLA-A*68:01~HLA-C*15:02~HLA-B*40:06+HLA-A*68:01~HLA-C*06:02~HLA-B*45:01
```
The "Gl String" column contains a [GL String](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3715123/pdf/tan0082-0106.pdf) formatted multi-locus HLA haplotype. In GL String format, the ~ operator denotes phase, and the + operator denotes copies of genes (in this case diploidy). While GL Strings can be used to describe ambiguous alleles and genotypes using the / and | operators, ambiguous data cannot be included in an *LDWrap()* data file. *LDWrap()* requires that all HLA alleles are written with full allele name prefixes (e.g., `HLA-DRB1*01:01`). HLA data written using short prefixes (`DRB1*01:01`) or no prefixes (`01:01`) are not allowed. Unusual allele names (`HLA-A*NULL`, `HLA-DRB1*NoMatch`, `HLA-DPB1*NT`) and truncated versions of allele names (`HLA-A*01`, `HLA-A*01:01`, `HLA-A*01:01:01`, etc.) will be analyzed as distinct alleles, and may skew analytic results.

### LD.sign.test

The *LD.sign.test()* function applies the R Stats Package's *binom.test()* function to pairs of LD values (*D'*, *Wn*, *WLoc1/Loc2*, *WLoc2/Loc1*), as well as the number of haplotypes, for phased and unphased haplotypes. The *_LD_results.csv* files generated by *LDWrap()* are the input files for this function, and generally, *LDWrap()* must be used before this function can be applied. See the *LDWrap()* Outputs section below for an example.

### LD.heat.map

The *LD.heat.map()* function generates PNG-formatted heat-maps of the LD values generated for each LD measure (*D'*, *Wn*, *WLoc1/Loc2*, and *WLoc2/Loc1*) for phased and unphased haplotpes. If LD values for only phased or unphased haplotpes are available, half-matrix heat-maps will be generated. The *_LD_results.csv* files generated by *LDWrap()* are the input files for this function, and generally, *LDWrap()* must be used before this function can be applied. See the *LDWrap()* Outputs section below for an example.

## Outputs

### cALD()

By default, *cALD()* operates in "verbose" mode, and will write five lines of output to the console describing the phase-status of the LD analysis (phased or unphased), the loci and number of haplotypes analyzed, and the four LD measures calculated, as shown below.
```
> cALD(drb1.dqb1.demo)
## Calculating D', Wn and conditional ALD for 53 unphased genotypes at the DRB1 and DQB1 loci.
## D' for DRB1~DQB1 haplotypes: 0.95892767844544 (0.9589) 
## Wn for DRB1~DQB1 haplotypes: 0.811250972337927 (0.8113) 
## Variation of DQB1 conditioned on DRB1 (WDQB1/DRB1) = 0.904035615838528 (0.904)
## Variation of DRB1 conditioned on DQB1 (WDRB1/DQB1) = 0.778712696009626 (0.7787)
```
When `verbose=FALSE`, *cALD()* returns a vector of *D<sup>'</sup>*, *W<sub>n</sub>*, *W<sub>B/A</sub>*, *W<sub>A/B</sub>* and the number of haplotypes, as below.
```
> cALDres <- cALD(drb1.dqb1.demo, verbose=FALSE)
> cALDres
## [1] "0.958463650196244" "0.811184752436694" "0.903300938910147" "0.778712697633606" "53"
```
In addition, *cALD()* can write a text file, containing a vector of all haplotypes, their frequencies and counts for the analyzed locus pair, to the working directory. This file also includes information on the dataset and phase status applied to the genotype data for the analysis. An example generted for the `drb1.dqb1.demo` dataset is shown below. 
```
> cALD(drb1.dqb1.demo,reportVector = TRUE)
```
Haplotype vector file contents:
```
Dataset	                                        Phase	DRB1~DQB1   Frequency	        Count
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	01:01~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	01:02~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	01:03~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	03:01~02:01	0.094272076372315	79
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	04:01~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	04:02~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	04:03~02:01	0	                0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12	FALSE	04:04~02:01	0	                0
.
.
.
```
### LDWrap
*LDWrap()* sends data to *cALD()*, captures the vector of LD results returned by *cALD()*, and directs *cALD()* to write vectors of haplotypes for each locus pair. As a single haplotype vector file is written for each locus pair, *LDWrap()* directs *cALD()* to write n(n-1)/2 haplotype vector files, where n is the number of loci in a haplotype. The only information *LDWrap()* returns to the console is a notification that the analysis has completed (`LD Analysis Complete`) and a notification that the input data is missing the required columns, as below.
```
> LDWrap(drb1.dqb1.demo)
LD Analysis Halted: Your data frame does not contain the proper columns. The 'Relation' and 'Gl String' columns are missing.
```
When all locus pairs have been analyzed by *cALD()*, *LDWrap()* writes a six-column CSV file (`*LD_results.csv`) of aggregated LD result vectors collected from *cALD()* to the working directory. The column headers in this file are Loc1~Loc2, identifying the locus pair, and  *D<sup>'</sup>*, *W<sub>n</sub>*, *W<sub>B/A</sub>*, *W<sub>A/B</sub>* and N_Haplotypes. An example of this file is shown below.

```
> LDWrap(hla.hap.demo)
LD Analysis Complete
```
LD results file contents: 
```
Loc1~Loc2,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
A~C,0.469024805013898,0.362566555750013,0.366359427624652,0.384413960789992,191
A~B,0.540780240853345,0.446662593270748,0.36839918931955,0.471334711300434,241
A~DRB1,0.400002012804198,0.335434108343871,0.27413544158564,0.320399398449896,233
A~DRB3,Not Calculated,Subject Threshold=10,Complete subjects=8,.,
.
.
.
```
*LDWrap()* attemtps to peform these LD calculations for all pairs of loci in the *LDWrap()* datafile. If a haplotype dataset includes locus pairs for which the number of subjects is below the `threshold` value (see "Parameters", below), the `*LD_results.csv` file will include rows for locus pairs for which no LD calculations were performed. As shown above, those rows contain data similar to that shown for the A~DRB3 haplotpe -- `Not Calculated`, `Subject Threshold=10`, `Complete subjects=8`, `.`, and ' '. In cases where at least one locus in a pair is monomorphic, no LD calculations are performed, and the pertinent rows in the `*LD_results.csv` file will contain the data, `Not Calculated`, `Subject Threshold=10`, `Complete subjects=0`, `"locusName" is monomorphic.`, and ``, as shown below.
```
Loc1~Loc2,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
B~DRB3,Not Calculated,Subject Threshold=10,Complete subjects=130,DRB3 is monomorphic.,
DRB3~DRB4,Not Calculated,Subject Threshold=10,Complete subjects=130,DRB3 is monomorphic. DRB4 is monomorphic.,
DRB3~DQA1,Not Calculated,Subject Threshold=10,Complete subjects=93,DRB3 is monomorphic.,
```
### LD.sign.test
For each LD measure and the number of haplotypes for phased and unphased versions of the same genotype data, *LD.sign.test()* reports the p-value of the sign test, comparing the number of locus pairs for which the value of the measure is higher in unphased haplotypes than phased haplotypes to the number of locus pairs for which that value is lower or equal. The function also reports the total number of locus pairs evaluated, and the number of locus pairs with equal values. These data can be reported in three ways; as a CSV file written to the working directory, as a table written to the console, and as a returned data frame. All three report formats are illustrated below. Note, only the significance of the sign test is reported; when a significant trend is indicated, the directionality of the trend is not reported.

```
> LD.sign.test("hla-family-data")
```
CSV File
```
,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
#unphased > phased,15,14,15,15,0
#unphased = phased,0,0,0,0,0
#locus pairs,15,15,15,15,15
p-values,6.10351562500001e-05,0.0009765625,6.10351562500001e-05,6.10351562500001e-05,6.10351562500001e-05
```
Console Table
```
Sign Test results for the hla-family-data dataset for 15 locus pairs.
Measure     	#U > P	#U = P	  p-value	
D'          	15	    0	      6.104e-05	
Wn          	14	    0	      0.0009766	
WLoc1/Loc2  	15	    0	      6.104e-05	
WLoc2/Loc1  	15	    0	      6.104e-05	
# Haplotypes	0	    0	      6.104e-05	
```
```
> LD.res <- LD.sign.test("hla-family-data",returnFrame = TRUE)
> View(LD.res)   
```
Returned Object
```
                             D'           Wn   WLoc1/Loc2   WLoc2/Loc1 N_Haplotypes
#unphased > phased 1.500000e+01 1.400000e+01 1.500000e+01 1.500000e+01 0.000000e+00
#unphased = phased 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
#locus pairs       1.500000e+01 1.500000e+01 1.500000e+01 1.500000e+01 1.500000e+01
p-values           6.103516e-05 9.765625e-04 6.103516e-05 6.103516e-05 6.103516e-05

```
### LD.heat.map
The *LD.heat.map()* function generates heat-map plots for each LD measure, visualizing the range of LD values for each locus-pair analzed using *LDWrap()*. LD values for phased data are presented in the upper half of the plot matrix, while values for unphased data are presnted in the lower half of the plot matrix. Color (blue to white to red) or greyscale (dark grey to light grey) plots can be specified. The heat-map plot for each LD measure is written to the working directory with the name `<dataSet>_<LD measure>_heatmap.png`. When no *LDwrap()* results are available, a notification is written to the console. There are no other outputs.

```
> LD.heat.map("family-data")
```
Console Notification
```
Files family-data_Phased_LD_results.csv and family-data_Unphased_LD_results.csv are not found.
```

## Parameters

### cALD
```
cALD(dataSet, inPhase = FALSE, verbose = TRUE, reportVector = FALSE, vectorName = "", vectorPrefix = "")
```

**dataSet**

Class: String. Required. No Default.

e.g., `dataSet="foo.txt"` or `dataSet=drb1.dqb1.demo`

Identifies the genotype data that you want to analyze. `dataSet` should identify a four-column data-frame or tab-delimited text file. Columns 1 and 2 contain genotype data for one genetic locus, with one allele in each column. Columns 3 and 4 contain genotype data for the second locus. The headers for columns 1 and 2 identify the first locus, and must be identical. Similarly, the headers for columns 3 and 4 identify the second locus, and must be identical.

**inPhase**

Class: Logical. Required. Default=FALSE.

Identifies how the genotype data should be analyzed. When `inPhase=FALSE`, the expectation-maximization (EM) algorithm is applied to the genotype data to generate estimated haplotypes. LD values are calculated for those EM haplotypes. When `inPhase=TRUE`, the genotype data are treated as phased, with the alleles in column 1 in phase with those in column 3, and the alleles in column 2 in phase with those in column 4. LD values are calculated for those phased haplotypes.

**verbose**

Class: Logical. Required. Default=TRUE.

Identifies how LD values should be reported. When `verbose=TRUE`, values for *D<sup>'</sup>*, *W<sub>n</sub>*, *W<sub>B/A</sub>*, *W<sub>A/B</sub>* and the number of haplotypes evaluated are written to the console in a human readable form. When `verbose=FALSE`, those values are reported in a vector of length 5 and mode of "character".

**reportVector**

Class: Logical. Required. Default=FALSE.

Specifies if a file containing the haplotype vector for the analyzed locus pair should be exported. If `reportVector=FALSE` no vector is written. If `reportVector=TRUE` a tab-delimited text file consisting of five columns is written to the working directory. The file includes the columns "Dataset", "Phase", the name of the analyzed locus pair, "Frequency", and "Count"; the latter two describe the frequency and count data for all possible haplptypes.

**vectorName**

Class: String. Optional. No Default. (`reportVector=TRUE` specific).

Provides a name for the tab-delimited haplotype vector file written when `reportVector=TRUE`. When `vectorName="foo"` and `reportVector=TRUE`, the name of the haplotype vector file will be "foo.txt". When `vectorName=""` and `reportVector=TRUE`, the name of the haplotype vector file will include the loci analyzed and a time-stamp, formatted as "Locus1~Locus2_haplotype_Vector_yyyy-MM-dd-HH-mm-ss.txt".

**vectorPrefix**

Class: String. Optional. No Default. (`reportVector=TRUE` specific).

Applies a prefix that includes the applied phase-status to the name of the tab-delimited haplotype vector file written when `reportVector=TRUE`, when  `vectorName=""`. If any string is provided for vectorName, this parameter is ignored. E.g., when `reportVector=TRUE`, `vectorName=""`, `vectorPrefix="foo"` and `inPhase=FALSE`, the haplotype vector file will be named "foo_unphased_Locus1~Locus2_haplotype_Vector_yyyy-MM-dd-HH-mm-ss.txt".

### LDWrap
```
LDWrap(famData, threshold = 10, phased = TRUE, frameName = "hla-family-data")
```

**famData**

Class: String. Required. No Default.

e.g., `famData="foo.csv"` or `famData=hla.hap.demo`

Identifies the haplotype dataset that you want to analyze. `famData` should identify a data-frame or CSV file. This dataset must inlcude columns with the headers "Relation" and "Gl String". If either (or both) of these column headers is not found in the dataset, or if the data file is not a CSV file, *LDWrap()* will halt the analysis with a notification about the missing header(s). See the Functions and Input Formats section above for additional details about the dataset format.

**threshold**

Class: Numeric. Required. Default=10.

Identifies the minimum number of subjects with haplotype data for a given locus pair required for the analysis of that locus pair. Analysis for that locus pair is not performed if the threshold is not met, and the LD results file will identify the threshold value and the number of subjects with data for that locus pair. If `threshold` is set to less than 1, it is automatically set to 1.

**phased**

Class: Logical. Required. Default=TRUE.

Specifies whether the haplotype data should be treated as phased (`phased=TRUE`) or unphased (`phased=FALSE`) for analysis.

**frameName**

Class: String. Optional. Default="hla-family-data".

Provides a name that will be included in the names of the result files if `famData` specifies a data frame. The value of `frameName` is passed to *cALD()* as the `vectorPrefix` parameter. If `famData` specifies a file, `frameName` is ignored.

### LD.sign.test
```
LD.sign.test(dataName,verbose = TRUE,returnFrame = FALSE)
```

**dataName** 

Class: String. Required. No Default.

e.g., `dataName="foo"`

The "base" name of the "_LD_result.csv" files generated by *LDWrap()*, with the "_Phased_LD_results.csv" or "_Unphased_LD_results.csv" suffixes removed. This corresponds to the value of the *LDWrap()* `frameName` parameter when the *LDWrap()* `famData` parameter does not specify a file; e.g., when specifying the "_LD_results.csv" files generated by *LDWrap()* for the `hla.hap.demo` data included with this package, `dataName="hla-family-data"`. If the corresponding "<dataName>_Phased_LD_results.csv" or "<dataName>_Unphased_LD_results.csv" files are not found, the function will halt with a notification identifying the file(s) that are not found. 

**verbose**

Class: Logical. Required. Default=TRUE.

Identifies if messages about function progress and results should be displayed in the console (verbose=TRUE) or not (verbose=FALSE). The default is verbose=TRUE. In addition to the table of results and messages about missing input files, messages regarding locus pairs for which no LD values are available, and about discrepancies between locus pairs with available data in the phased and unphased datasets are also suppressed with verbose=FALSE.

**returnFrame**

Class: Logical. Required. Default=FALSE.

Identifies if a data frame of results should be returned (returnFrame=TRUE). If returnFrame=FALSE, a CSV file of results named "<dataName>_LD-sign-test_results.csv"" is written in the working directory.

### LD.heat.map
```
LD.heat.map(dataName)
```

**dataName** 

Class: String. Required. No Default.

e.g., `dataName="foo"`

The "base" name of the "_LD_result.csv" files generated by *LDWrap()*, with the "_Phased_LD_results.csv" or "_Unphased_LD_results.csv" suffixes removed. This corresponds to the value of the *LDWrap()* `frameName` parameter when the *LDWrap()* `famData` parameter does not specify a file; e.g., when specifying the "_LD_results.csv" files generated by *LDWrap()* for the `hla.hap.demo` data included with this package, `dataName="hla-family-data"`. If the corresponding "<dataName>_Phased_LD_results.csv" and "<dataName>_Unphased_LD_results.csv" files are not found, the function will halt with a notification identifying the files that are not found. 

**color**

Class: Logical. Required. Default=TRUE.

Identifies if the heat-map plots should be generated in color (color=TRUE) or greyscale (color=FALSE). The default is color=TRUE. Color heat-map plots will range from blue (low LD values) to white (LD values of 0.5) to red (high LD values). Greyscale heat-map plots will range from dark grey (Low LD values) to light grey (high LD values).

**phasedLabel**

Class: String. Required. Default="Phased".

e.g., `phasedLabel="Pedigree Phased"`

Specifies the label that should appear on the heat-map plots for the upper, phased half of the plot.

**unphasedLabel**

Class: String. Required. Default="EM-estimated".

e.g., `phasedLabel="EM Haplotypes"`

Specifies the label that should appear on the heat-map plots for the lower, unphased half of the plot.

## Examples
###cALD()
```
# Analyzing the included HLA-DRB1 HLA-DQB1 genotype data and reporting results in the console
cALD(drb1.dqb1.demo)
# Alternatively returning a vector of LD results
LDvec <- cALD(drb1.dqb1.demo,verbose=FALSE)
LDvec
# Enforcing phase between columns 1 and 3 and between columns 2 and 4 for analysis
cALD(drb1.dqb1.demo,inPhase=TRUE)
# Writing the haplotype vector to a file in the working directory that will 
# have the time-stamped name DRB1~DQB1_haplotype_Vector_yyyy-MM-dd_HH-mm-ss.txt
cALD(drb1.dqb1.demo,reportVector = TRUE)
# Writing the haplotype vector to a file in the working directory named "foo.txt" 
cALD(drb1.dqb1.demo,reportVector = TRUE,vectorName = "foo")
# Writing the haplotype vector to a file in the working directory with the prefix "foo_"
cALD(drb1.dqb1.demo,reportVector = TRUE,vectorPrefix = "foo")
```

###LDWrap()
```
# Analyzing the included HLA haplotype data 
# This will create 15 haplotype vector files and one LD results file in the working directory
LDWrap(hla.hap.demo)
# Specifying the prefix "foo_Phased" for the LD results file, and "foo_phased" for the haplotype vector files
LDWrap(hla.hap.demo,frameName = "foo")
```

###LD.sign.test()
```
# Generating LDWrap() results files for the example data included with this package.
LDWrap(hla.hap.demo)
LDWrap(hla.hap.demo,phased=FALSE)
# Analyzing the results files generated by LDWrap().
LD.sign.test("hla-family-data")
# Returning only a data frame for the same analysis.
LD.res <- LD.sign.test("hla-family-data",verbose=FALSE,returnFrame = TRUE)
View(LD.res)
```

###LD.heat.map()
```
# Generating LDWrap() results files for the example data included with this package.
LDWrap(hla.hap.demo)
LDWrap(hla.hap.demo,phased=FALSE)
# Generating color heat-map plots in the working directory.
LD.heat.map("hla-family-data")
# Generating greyscale heat-map plots in the working directory.
LD.heat.map("hla-family-data",color=FALSE)

```
##
End of vignette.
